python_modules/dagster/dagster/core/asset_defs/asset_selection.py

```diff
@@ -0,0 +1,142 @@
+import operator
+from abc import ABC
+from functools import reduce
+from typing import AbstractSet, FrozenSet, Optional
+
+from typing_extensions import TypeAlias
+
+import dagster._check as check
+from dagster.core.asset_defs.assets import AssetsDefinition
+from dagster.core.selector.subset_selector import (
+    Direction,
+    fetch_connected_assets_definitions,
+    generate_asset_dep_graph,
+    generate_asset_name_to_definition_map,
+)
+
+AssetSet: TypeAlias = AbstractSet[AssetsDefinition]  # makes sigs more readable
+
+
+class AssetSelection(ABC):
+    @staticmethod
+    def all() -> "AllAssetSelection":
+        return AllAssetSelection()
+
+    @staticmethod
+    def keys(*key_strs: str) -> "KeysAssetSelection":
+        return KeysAssetSelection(*key_strs)
+
+    @staticmethod
+    def groups(*group_strs) -> "GroupsAssetSelection":
+        return GroupsAssetSelection(*group_strs)
+
+    def downstream(self, depth: Optional[int] = None) -> "DownstreamAssetSelection":
+        return DownstreamAssetSelection(self, depth=depth)
+
+    def upstream(self, depth: Optional[int] = None) -> "UpstreamAssetSelection":
+        return UpstreamAssetSelection(self, depth=depth)
+
+    def __or__(self, other: "AssetSelection") -> "OrAssetSelection":
+        return OrAssetSelection(self, other)
+
+    def __and__(self, other: "AssetSelection") -> "AndAssetSelection":
+        return AndAssetSelection(self, other)
+
+    def resolve(self, all_assets: AssetSet) -> AssetSet:
+        return Resolver(all_assets).resolve(self)
+
+
+class AllAssetSelection(AssetSelection):
+    pass
+
+
+class AndAssetSelection(AssetSelection):
+    def __init__(self, child_1: AssetSelection, child_2: AssetSelection):
+        self.children = (child_1, child_2)
+
+
+class DownstreamAssetSelection(AssetSelection):
+    def __init__(self, child: AssetSelection, *, depth: Optional[int] = None):
+        self.children = (child,)
+        self.depth = depth
+
+
+class GroupsAssetSelection(AssetSelection):
+    def __init__(self, *children: str):
+        self.children = children
+
+
+class KeysAssetSelection(AssetSelection):
+    def __init__(self, *children: str):
+        self.children = children
+
+
+class OrAssetSelection(AssetSelection):
+    def __init__(self, child_1: AssetSelection, child_2: AssetSelection):
+        self.children = (child_1, child_2)
+
+
+class UpstreamAssetSelection(AssetSelection):
+    def __init__(self, child: AssetSelection, *, depth: Optional[int] = None):
+        self.children = (child,)
+        self.depth = depth
+
+
+# ########################
+# ##### RESOLUTION
+# ########################
+
+
+class Resolver:
```

Comment:
Thoughts on making this a function instead of a class? That would make it crystal clear what data `resolve` operates on (because it would be in the signature).

Reply:
It's a class because it needs to essentially cache some intermediate data structures (dependency graph and asset lookup table) that are used for the duration of the computation, which includes multiple recursive node resolutions.